### PR TITLE
Add support for handling RPM databases in NDB format

### PR DESCRIPTION
**Pull request by dirkmueller**
RPM databases can be either in BerkeleyDB format or optionally in the
newer "native" NDB format. there are linux distributions that switched
their implementation of the rpm database away from BerkeleyDB out of
concerns about future level of maintenance after its license change.

See https://rpm.org/wiki/Releases/4.14.0
See also https://fedoraproject.org/wiki/Changes/NewRpmDBFormat

This adds a simple interface that allows to drive either berkelydb
or NDB format reading.

*Issue #, if available:*
https://github.com/aquasecurity/fanal/issues/183

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/ndb/ndb.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+			if err != nil {
+				entries <- dbi.Entry{
+					Err: err,
+				}
+				return
+			}
+			const NDB_BlobMagic = 'B' | 'l'<<8 | 'b'<<16 | 'S'<<24
+			if blobHeaderBuff.BlobMagic != NDB_BlobMagic {
+				entries <- dbi.Entry{
+					Err: xerrors.Errorf("Unexpected NDB blob Magic for pkg %d: %x", slot.PkgIndex, blobHeaderBuff.BlobMagic),
```
Comment (knqyf263) [inline pkg/ndb/ndb.go:131]: nit:
```suggestion
					Err: xerrors.Errorf("unexpected NDB blob Magic for pkg %d: %x", slot.PkgIndex, blobHeaderBuff.BlobMagic),
```

File: pkg/ndb/ndb.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+				}
+			}
+			if blobHeaderBuff.PkgIndex != slot.PkgIndex {
+				entries <- dbi.Entry{
+					Err: xerrors.Errorf("Failed to find NDB blob for pkg %d", slot.PkgIndex),
```
Comment (knqyf263) [inline pkg/ndb/ndb.go:136]: nit:
```suggestion
					Err: xerrors.Errorf("failed to find NDB blob for pkg %d", slot.PkgIndex),
```

File: pkg/rpmdb_testcase.go [hunk 3]
```diff
@@ -2785,6 +2785,46 @@ var (
 		{0, "ostree-libs", "2019.6", "2.el8", "x86_64", "ostree-2019.6-2.el8.src.rpm", 981292, "LGPLv2+", "CentOS", "", nil, nil, nil},
 	}
 
+	// docker run --rm -it registry.suse.com/bci/minimal:15.3 bash
+	// rpm -qa --queryformat "\{%{EPOCH}, \"%{NAME}\", \"%{VERSION}\", \"%{RELEASE}\", \"%{ARCH}\"\, \"%{SOURCERPM}\", %{SIZE}, \"%{LICENSE}\", \"%{VENDOR}\", \"\", nil, nil, nil\},\n" | sed "s/(none)/0/g"
+	SLE15_3_bci = []*PackageInfo{
```
Comment (knqyf263) [inline pkg/rpmdb_testcase.go:2790]: Can we clarify this case is for NDB? Like SLE15WithNDB or something like that.

File: pkg/ndb/ndb.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+	hdrBuff := ndbHeader{}
+	err = binary.Read(file, binary.LittleEndian, &hdrBuff)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to read metadata: %w", err)
+	}
+
+	if hdrBuff.HeaderMagic != NDB_HeaderMagic || hdrBuff.SlotNPages == 0 {
+		// This does not look like a NDB, return without error
+		return nil, nil
```
Comment (knqyf263) [inline pkg/ndb/ndb.go:78]: Can we define a custom error? Like

```
var ErrInvalidNDB = errors.New("invalid NDB format")

File: pkg/rpmdb.go [hunk 5]
```diff
@@ -2,21 +2,33 @@ package rpmdb
...
-	db *bdb.BerkeleyDB
+	db dbi.RpmDBInterface
 }
 
 func Open(path string) (*RpmDB, error) {
-	db, err := bdb.Open(path)
+
+	// NDB Open() returns nil, nil in case of DB format other than NDB
+	ndb, err := ndb.Open(path)
+	if err != nil {
```
Comment (knqyf263) [inline pkg/rpmdb.go:18]:
```suggestion
	if err != nil && !errors.Is(err, ndb.ErrInvalidNDB) {
```

File: pkg/ndb/ndb.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+					Err: xerrors.Errorf("Bad slot Magic: %x", slot.SlotMagic),
+				}
+				return
+			}
+			// Empty slot?
+			if slot.PkgIndex == 0 {
+				continue
+			}
+			blobHeaderBuff := ndbBlobHeader{}
+			seekOffset := int64(slot.BlkOffset) * int64(unsafe.Sizeof(blobHeaderBuff))
```
Comment (knqyf263) [inline pkg/ndb/ndb.go:115]: Do we have to calculate the size of `ndbBlobHeader` every time? Or, can we do it once outside of this function?

**Review by knqyf263 — COMMENTED** [review]
Looks great!! Thanks a lot! I left some small comments.

BTW, how did you get the specification of NDB? I didn't find it. If you referred to the source code of `rpm`, could you add the reference as a comment like the following?
https://github.com/knqyf263/go-rpmdb/blob/master/pkg/entry.go#L13

File: pkg/rpmdb_testcase.go (hunk 3 above)
Comment (dirkmueller) [reply pkg/rpmdb_testcase.go:2790]: absolutely, done.

File: pkg/ndb/ndb.go (hunk 6 above)
Comment (dirkmueller) [reply pkg/ndb/ndb.go:115]: Ah, I wasn't assuming that this is cost intensive. the size is fixed to 32 bytes, I can just define a constant for it?

Comment (dirkmueller) [reply pkg/ndb/ndb.go:115]: done.

**Comment by dirkmueller** [comment]
> BTW, how did you get the specification of NDB? I didn't find it. If you referred to the source code of `rpm`, could you add the reference as a comment like the following? https://github.com/knqyf263/go-rpmdb/blob/master/pkg/entry.go#L13

My colleague is the original author of NDB, and I asked him to get a suitable DB dumping tool under a non-GPL license  [which I attached to the fanal issue](https://github.com/aquasecurity/fanal/issues/183#issuecomment-861435228)

I have added a brief comment with the format description into the ndb implementation here.

File: pkg/ndb/ndb.go (hunk 4 above)
Comment (dirkmueller) [reply pkg/ndb/ndb.go:78]: sure. done

File: pkg/rpmdb.go (hunk 5 above)
Comment (dirkmueller) [reply pkg/rpmdb.go:18]: implemented, thanks for the suggestion!

**Comment by dirkmueller** [comment]
@knqyf263 anything else I can do for you?

File: pkg/ndb/ndb.go (hunk 6 above)
Comment (knqyf263) [reply pkg/ndb/ndb.go]: Thanks. It's no big deal actually, but it might be better to avoid doing that many times if possible.

**Review by knqyf263 — APPROVED** [review]
LGTM :shipit:

**Comment by knqyf263** [comment]
Thanks again for your great contribution!

**Pull request merged**